<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>12 Lecture 11 - Sequence analysis in R II | [2019 UST 일반강의] R 프로그래밍과 데이터 분석</title>
  <meta name="description" content="2019년 UST 강의 R 프로그래밍 웹사이트">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="12 Lecture 11 - Sequence analysis in R II | [2019 UST 일반강의] R 프로그래밍과 데이터 분석" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="2019년 UST 강의 R 프로그래밍 웹사이트" />
  <meta name="github-repo" content="https://greendaygh.github.io/Rprog2019/" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="12 Lecture 11 - Sequence analysis in R II | [2019 UST 일반강의] R 프로그래밍과 데이터 분석" />
  
  <meta name="twitter:description" content="2019년 UST 강의 R 프로그래밍 웹사이트" />
  

<meta name="author" content="Haseong Kim">


<meta name="date" content="2019-06-03">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="lecture-10-sequence-analysis-in-r-i.html">

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Lecture 개요</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#goal--"><i class="fa fa-check"></i><b>1.1</b> Goal 강의 목표</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#references--"><i class="fa fa-check"></i><b>1.2</b> References 참고 자료</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#evaluation---"><i class="fa fa-check"></i><b>1.3</b> Evaluation 평가 세부 항목</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#schedule--"><i class="fa fa-check"></i><b>1.4</b> Schedule 강의 계획</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#lecture-note--"><i class="fa fa-check"></i><b>1.5</b> Lecture note 강의 자료</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="lecture-01-r-introduction.html"><a href="lecture-01-r-introduction.html"><i class="fa fa-check"></i><b>2</b> Lecture 01 - R Introduction</a></li>
<li class="chapter" data-level="3" data-path="lecture-02-r-basics-i.html"><a href="lecture-02-r-basics-i.html"><i class="fa fa-check"></i><b>3</b> Lecture 02 - R basics I</a><ul>
<li class="chapter" data-level="3.1" data-path="lecture-02-r-basics-i.html"><a href="lecture-02-r-basics-i.html#exercise-2-1-who-function"><i class="fa fa-check"></i><b>3.1</b> exercise 2-1: who function</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="lecture-03-r-basics-ii.html"><a href="lecture-03-r-basics-ii.html"><i class="fa fa-check"></i><b>4</b> Lecture 03 - R Basics II</a><ul>
<li class="chapter" data-level="4.1" data-path="lecture-03-r-basics-ii.html"><a href="lecture-03-r-basics-ii.html#matrix-"><i class="fa fa-check"></i><b>4.1</b> matrix (행렬)</a></li>
<li class="chapter" data-level="4.2" data-path="lecture-03-r-basics-ii.html"><a href="lecture-03-r-basics-ii.html#data.frame"><i class="fa fa-check"></i><b>4.2</b> data.frame</a></li>
<li class="chapter" data-level="4.3" data-path="lecture-03-r-basics-ii.html"><a href="lecture-03-r-basics-ii.html#list"><i class="fa fa-check"></i><b>4.3</b> list</a></li>
<li class="chapter" data-level="4.4" data-path="lecture-03-r-basics-ii.html"><a href="lecture-03-r-basics-ii.html#file-"><i class="fa fa-check"></i><b>4.4</b> File 읽기/쓰기</a></li>
<li class="chapter" data-level="4.5" data-path="lecture-03-r-basics-ii.html"><a href="lecture-03-r-basics-ii.html#excel-file-"><i class="fa fa-check"></i><b>4.5</b> Excel file 읽기/쓰기</a></li>
<li class="chapter" data-level="4.6" data-path="lecture-03-r-basics-ii.html"><a href="lecture-03-r-basics-ii.html#exercise-3-2-standard-deviation"><i class="fa fa-check"></i><b>4.6</b> Exercise 3-2 standard deviation</a></li>
<li class="chapter" data-level="4.7" data-path="lecture-03-r-basics-ii.html"><a href="lecture-03-r-basics-ii.html#exercise-3-3-barplot-with-error-bars"><i class="fa fa-check"></i><b>4.7</b> Exercise 3-3 barplot with error bars</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="lecture-04-maze-robot.html"><a href="lecture-04-maze-robot.html"><i class="fa fa-check"></i><b>5</b> Lecture 04 - Maze Robot</a><ul>
<li class="chapter" data-level="5.1" data-path="lecture-04-maze-robot.html"><a href="lecture-04-maze-robot.html#parameterargument-globallocal"><i class="fa fa-check"></i><b>5.1</b> Parameter/Argument, Global/Local</a></li>
<li class="chapter" data-level="5.2" data-path="lecture-04-maze-robot.html"><a href="lecture-04-maze-robot.html#exercise-4-1-matrix-indexing"><i class="fa fa-check"></i><b>5.2</b> Exercise 4-1: Matrix indexing</a></li>
<li class="chapter" data-level="5.3" data-path="lecture-04-maze-robot.html"><a href="lecture-04-maze-robot.html#top-down-design"><i class="fa fa-check"></i><b>5.3</b> Top-down design</a></li>
<li class="chapter" data-level="5.4" data-path="lecture-04-maze-robot.html"><a href="lecture-04-maze-robot.html#exercise-4-2-draw-world"><i class="fa fa-check"></i><b>5.4</b> Exercise 4-2: Draw world</a></li>
<li class="chapter" data-level="5.5" data-path="lecture-04-maze-robot.html"><a href="lecture-04-maze-robot.html#exercise-4-4-generate-map"><i class="fa fa-check"></i><b>5.5</b> Exercise 4-4: Generate map</a></li>
<li class="chapter" data-level="5.6" data-path="lecture-04-maze-robot.html"><a href="lecture-04-maze-robot.html#draw-blocks-and-walls"><i class="fa fa-check"></i><b>5.6</b> Draw blocks and walls</a></li>
<li class="chapter" data-level="5.7" data-path="lecture-04-maze-robot.html"><a href="lecture-04-maze-robot.html#move-one-step"><i class="fa fa-check"></i><b>5.7</b> Move one step</a></li>
<li class="chapter" data-level="5.8" data-path="lecture-04-maze-robot.html"><a href="lecture-04-maze-robot.html#sleep-for-delay"><i class="fa fa-check"></i><b>5.8</b> Sleep for delay</a></li>
<li class="chapter" data-level="5.9" data-path="lecture-04-maze-robot.html"><a href="lecture-04-maze-robot.html#move-50-steps"><i class="fa fa-check"></i><b>5.9</b> Move 50 steps</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="lecture-05-multiple-plate-reader-file.html"><a href="lecture-05-multiple-plate-reader-file.html"><i class="fa fa-check"></i><b>6</b> Lecture 05 - Multiple plate reader file</a><ul>
<li class="chapter" data-level="6.1" data-path="lecture-05-multiple-plate-reader-file.html"><a href="lecture-05-multiple-plate-reader-file.html#file-io--"><i class="fa fa-check"></i><b>6.1</b> File I/O 파일 입출력</a></li>
<li class="chapter" data-level="6.2" data-path="lecture-05-multiple-plate-reader-file.html"><a href="lecture-05-multiple-plate-reader-file.html#plot-"><i class="fa fa-check"></i><b>6.2</b> plot 산포도</a></li>
<li class="chapter" data-level="6.3" data-path="lecture-05-multiple-plate-reader-file.html"><a href="lecture-05-multiple-plate-reader-file.html#hist-"><i class="fa fa-check"></i><b>6.3</b> hist 히스토그램</a></li>
<li class="chapter" data-level="6.4" data-path="lecture-05-multiple-plate-reader-file.html"><a href="lecture-05-multiple-plate-reader-file.html#boxplot-"><i class="fa fa-check"></i><b>6.4</b> boxplot 상자그래프</a></li>
<li class="chapter" data-level="6.5" data-path="lecture-05-multiple-plate-reader-file.html"><a href="lecture-05-multiple-plate-reader-file.html#barplot-"><i class="fa fa-check"></i><b>6.5</b> barplot 막대그래프</a></li>
<li class="chapter" data-level="6.6" data-path="lecture-05-multiple-plate-reader-file.html"><a href="lecture-05-multiple-plate-reader-file.html#other-examples--"><i class="fa fa-check"></i><b>6.6</b> other examples 그래프 예시</a></li>
<li class="chapter" data-level="6.7" data-path="lecture-05-multiple-plate-reader-file.html"><a href="lecture-05-multiple-plate-reader-file.html#multiplate-reader--"><i class="fa fa-check"></i><b>6.7</b> Multiplate reader 기능 구현</a></li>
<li class="chapter" data-level="6.8" data-path="lecture-05-multiple-plate-reader-file.html"><a href="lecture-05-multiple-plate-reader-file.html#multiplate-reader-function-"><i class="fa fa-check"></i><b>6.8</b> Multiplate reader function 만들기</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="lecture-06-multiple-plate-reader-ggplot.html"><a href="lecture-06-multiple-plate-reader-ggplot.html"><i class="fa fa-check"></i><b>7</b> Lecture 06 - Multiple plate reader / ggplot</a><ul>
<li class="chapter" data-level="7.1" data-path="lecture-06-multiple-plate-reader-ggplot.html"><a href="lecture-06-multiple-plate-reader-ggplot.html#multiple_plate_excel_reader--"><i class="fa fa-check"></i><b>7.1</b> multiple_plate_excel_reader 함수 만들기</a></li>
<li class="chapter" data-level="7.2" data-path="lecture-06-multiple-plate-reader-ggplot.html"><a href="lecture-06-multiple-plate-reader-ggplot.html#multiple_plate_excel_reader2--"><i class="fa fa-check"></i><b>7.2</b> multiple_plate_excel_reader2 함수 업데이트</a></li>
<li class="chapter" data-level="7.3" data-path="lecture-06-multiple-plate-reader-ggplot.html"><a href="lecture-06-multiple-plate-reader-ggplot.html#read-data"><i class="fa fa-check"></i><b>7.3</b> Read data</a></li>
<li class="chapter" data-level="7.4" data-path="lecture-06-multiple-plate-reader-ggplot.html"><a href="lecture-06-multiple-plate-reader-ggplot.html#ggplot--barplot-"><i class="fa fa-check"></i><b>7.4</b> ggplot 이용 barplot 그리기</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="lecture-07-ggplot-data-transformation.html"><a href="lecture-07-ggplot-data-transformation.html"><i class="fa fa-check"></i><b>8</b> Lecture 07 - ggplot / data transformation</a><ul>
<li class="chapter" data-level="8.1" data-path="lecture-07-ggplot-data-transformation.html"><a href="lecture-07-ggplot-data-transformation.html#bar-graph-with-basic-graph-functions"><i class="fa fa-check"></i><b>8.1</b> Bar graph with basic graph functions</a></li>
<li class="chapter" data-level="8.2" data-path="lecture-07-ggplot-data-transformation.html"><a href="lecture-07-ggplot-data-transformation.html#bar-graph-with-ggplot"><i class="fa fa-check"></i><b>8.2</b> Bar graph with ggplot</a></li>
<li class="chapter" data-level="8.3" data-path="lecture-07-ggplot-data-transformation.html"><a href="lecture-07-ggplot-data-transformation.html#line-graph-with-ggplot"><i class="fa fa-check"></i><b>8.3</b> Line graph with ggplot</a></li>
<li class="chapter" data-level="8.4" data-path="lecture-07-ggplot-data-transformation.html"><a href="lecture-07-ggplot-data-transformation.html#bar-graph-of-the-example-dataset"><i class="fa fa-check"></i><b>8.4</b> Bar graph of the example dataset</a></li>
<li class="chapter" data-level="8.5" data-path="lecture-07-ggplot-data-transformation.html"><a href="lecture-07-ggplot-data-transformation.html#ddplyr---pipe-operator"><i class="fa fa-check"></i><b>8.5</b> ddplyr - pipe operator</a></li>
<li class="chapter" data-level="8.6" data-path="lecture-07-ggplot-data-transformation.html"><a href="lecture-07-ggplot-data-transformation.html#dplyr---important-functions"><i class="fa fa-check"></i><b>8.6</b> dplyr - Important functions</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="lecture-08-ggplot-data-transformation-using-dplyr.html"><a href="lecture-08-ggplot-data-transformation-using-dplyr.html"><i class="fa fa-check"></i><b>9</b> Lecture 08 - ggplot / data transformation using dplyr</a><ul>
<li class="chapter" data-level="9.1" data-path="lecture-08-ggplot-data-transformation-using-dplyr.html"><a href="lecture-08-ggplot-data-transformation-using-dplyr.html#dplyr-mutate-select-"><i class="fa fa-check"></i><b>9.1</b> dplyr mutate / select 함수</a></li>
<li class="chapter" data-level="9.2" data-path="lecture-08-ggplot-data-transformation-using-dplyr.html"><a href="lecture-08-ggplot-data-transformation-using-dplyr.html#ggplot2----bar-graph--"><i class="fa fa-check"></i><b>9.2</b> ggplot2를 이용한 두 개의 bar graph 겹쳐 그리기</a></li>
<li class="chapter" data-level="9.3" data-path="lecture-08-ggplot-data-transformation-using-dplyr.html"><a href="lecture-08-ggplot-data-transformation-using-dplyr.html#dplyr--iris-data-"><i class="fa fa-check"></i><b>9.3</b> dplyr 이용한 iris data 가공</a></li>
<li class="chapter" data-level="9.4" data-path="lecture-08-ggplot-data-transformation-using-dplyr.html"><a href="lecture-08-ggplot-data-transformation-using-dplyr.html#ggplot--iris-data-------"><i class="fa fa-check"></i><b>9.4</b> ggplot을 이용한 iris data 평균 값 및 에러 비교 그래프 그리기</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="lecture-09-ggplot-dlyr-for-96well-fluorescence-data.html"><a href="lecture-09-ggplot-dlyr-for-96well-fluorescence-data.html"><i class="fa fa-check"></i><b>10</b> Lecture 09 - ggplot / dlyr for 96well fluorescence data</a><ul>
<li class="chapter" data-level="10.1" data-path="lecture-09-ggplot-dlyr-for-96well-fluorescence-data.html"><a href="lecture-09-ggplot-dlyr-for-96well-fluorescence-data.html#dplyr-functions"><i class="fa fa-check"></i><b>10.1</b> dplyr functions</a></li>
<li class="chapter" data-level="10.2" data-path="lecture-09-ggplot-dlyr-for-96well-fluorescence-data.html"><a href="lecture-09-ggplot-dlyr-for-96well-fluorescence-data.html#data-transformation-using-melt-function"><i class="fa fa-check"></i><b>10.2</b> Data transformation using melt function</a></li>
<li class="chapter" data-level="10.3" data-path="lecture-09-ggplot-dlyr-for-96well-fluorescence-data.html"><a href="lecture-09-ggplot-dlyr-for-96well-fluorescence-data.html#scale-and-theme-of-ggplot2"><i class="fa fa-check"></i><b>10.3</b> scale and theme of ggplot2</a></li>
<li class="chapter" data-level="10.4" data-path="lecture-09-ggplot-dlyr-for-96well-fluorescence-data.html"><a href="lecture-09-ggplot-dlyr-for-96well-fluorescence-data.html#bargraph-for-96well-dataset"><i class="fa fa-check"></i><b>10.4</b> bargraph for 96well dataset</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="lecture-10-sequence-analysis-in-r-i.html"><a href="lecture-10-sequence-analysis-in-r-i.html"><i class="fa fa-check"></i><b>11</b> Lecture 10 - Sequence analysis in R I</a><ul>
<li class="chapter" data-level="11.1" data-path="lecture-10-sequence-analysis-in-r-i.html"><a href="lecture-10-sequence-analysis-in-r-i.html#biostrings-package-for-sequence-analysis"><i class="fa fa-check"></i><b>11.1</b> Biostrings package for sequence analysis</a><ul>
<li class="chapter" data-level="11.1.1" data-path="lecture-10-sequence-analysis-in-r-i.html"><a href="lecture-10-sequence-analysis-in-r-i.html#exercise-10-1"><i class="fa fa-check"></i><b>11.1.1</b> Exercise 10-1</a></li>
<li class="chapter" data-level="11.1.2" data-path="lecture-10-sequence-analysis-in-r-i.html"><a href="lecture-10-sequence-analysis-in-r-i.html#exercise-10-2"><i class="fa fa-check"></i><b>11.1.2</b> Exercise 10-2</a></li>
<li class="chapter" data-level="11.1.3" data-path="lecture-10-sequence-analysis-in-r-i.html"><a href="lecture-10-sequence-analysis-in-r-i.html#excercise-10-3"><i class="fa fa-check"></i><b>11.1.3</b> Excercise 10-3</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="lecture-10-sequence-analysis-in-r-i.html"><a href="lecture-10-sequence-analysis-in-r-i.html#download-sequence-information-from-ncbi"><i class="fa fa-check"></i><b>11.2</b> Download sequence information from NCBI</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="lecture-11-sequence-analysis-in-r-ii.html"><a href="lecture-11-sequence-analysis-in-r-ii.html"><i class="fa fa-check"></i><b>12</b> Lecture 11 - Sequence analysis in R II</a><ul>
<li class="chapter" data-level="12.1" data-path="lecture-11-sequence-analysis-in-r-ii.html"><a href="lecture-11-sequence-analysis-in-r-ii.html#genbank-data-parsing"><i class="fa fa-check"></i><b>12.1</b> Genbank data parsing</a><ul>
<li class="chapter" data-level="12.1.1" data-path="lecture-11-sequence-analysis-in-r-ii.html"><a href="lecture-11-sequence-analysis-in-r-ii.html#exercise-11-1"><i class="fa fa-check"></i><b>12.1.1</b> Exercise 11-1</a></li>
</ul></li>
<li class="chapter" data-level="12.2" data-path="lecture-11-sequence-analysis-in-r-ii.html"><a href="lecture-11-sequence-analysis-in-r-ii.html#iranges-for-genome-view"><i class="fa fa-check"></i><b>12.2</b> IRanges for genome view</a><ul>
<li class="chapter" data-level="12.2.1" data-path="lecture-11-sequence-analysis-in-r-ii.html"><a href="lecture-11-sequence-analysis-in-r-ii.html#exercise-11-2"><i class="fa fa-check"></i><b>12.2.1</b> Exercise 11-2</a></li>
<li class="chapter" data-level="12.2.2" data-path="lecture-11-sequence-analysis-in-r-ii.html"><a href="lecture-11-sequence-analysis-in-r-ii.html#exercise-11-3"><i class="fa fa-check"></i><b>12.2.2</b> Exercise 11-3</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="lecture-11-sequence-analysis-in-r-ii.html"><a href="lecture-11-sequence-analysis-in-r-ii.html#feature-view-using-iranges-and-ggplot2"><i class="fa fa-check"></i><b>12.3</b> Feature view using IRanges and ggplot2</a><ul>
<li class="chapter" data-level="12.3.1" data-path="lecture-11-sequence-analysis-in-r-ii.html"><a href="lecture-11-sequence-analysis-in-r-ii.html#exercise-11-4"><i class="fa fa-check"></i><b>12.3.1</b> Exercise 11-4</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">[2019 UST 일반강의] R 프로그래밍과 데이터 분석</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="lecture-11---sequence-analysis-in-r-ii" class="section level1">
<h1><span class="header-section-number">12</span> Lecture 11 - Sequence analysis in R II</h1>
<p>강의 노트 <a href="https://github.com/greendaygh/2019-R-programming/blob/master/11-Rprog.pdf">Lecture note</a></p>
<p>R을 이용한 서열분석 두 번째 시간입니다. 지난시간 NCBI 데이터를 활용하기 위한 reutils에 대한 내용에 이어서 서열 데이터를 다루기 위한 방법들을 배우도록 하겠습니다. 또한 예제에서 나오는 프로그래밍에 익숙해지기 위한 <code>for</code>, <code>if</code>, 그리고 함수를 만드는 법에 대한 내용도 참고하시기 바랍니다.</p>
<p>설치 필요 패키지 (사전 설치 필요). 설치는 다음 패키지들을 CRAN (RStudio 패키지 설치) 또는 Bioconductor 사이트에서 검색 후 지시에 따라 설치하시면 되겠습니다.</p>
<ol style="list-style-type: decimal">
<li>genbankr<br />
</li>
<li>GenomicRanges<br />
</li>
<li>ggbio<br />
</li>
<li>msa<br />
</li>
<li>seqinr<br />
</li>
<li>ape</li>
</ol>
<p>뎅기바이러스 데이터 다운로드</p>
<p><a href="https://github.com/greendaygh/2019-R-programming/blob/master/NC_001477.gb">NC_001477</a><br />
<a href="https://github.com/greendaygh/2019-R-programming/blob/master/NC_001474.gb">NC_001474</a><br />
<a href="https://github.com/greendaygh/2019-R-programming/blob/master/NC_001475.gb">NC_001475</a><br />
<a href="https://github.com/greendaygh/2019-R-programming/blob/master/NC_002640.gb">NC_002640</a></p>
<p>지난시간에는 reutils 패키지를 이용해 fasta format의 뎅기바이러스 데이터를 다운로드 받았습니다. 이번에는 gb (genbank) format의 데이터를 받아보도록 하겠습니다. genbank 포맷의 데이터는 서열 정보 뿐만 아니라 서열에 위치한 CDS 등 유전자 및 전사된 단백질 정보를 포함하고 있습니다.</p>
<p>다음 코드로 뎅기바이러스 4종에 대한 genbank 포맷의 데이터를 다운로드 할 수 있습니다. 앞서와 같이 4종에 대한 accession 번호를 알고 있다고 가정합니다. 중요한 부분은 <code>efetch</code> 입니다. 앞서 9개의 주요 함수들 중 특정 정보를 찾을때는 <code>esearch</code>를 사용하고 데이터를 다운로드 받을 때는 <code>efetch</code>를 사용한다고 말씀 드렸습니다. 지난시간 보았던 fasta format 다운로드 예제와 같이 <code>efetch</code>의 첫 번째와 두 번째 파라메터는 <code>epost</code> 함수를 사용하여 그 리턴값으로 대신할 수도 있습니다.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(reutils)

acc &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;NC_001477&quot;</span>, <span class="st">&quot;NC_001474&quot;</span>, <span class="st">&quot;NC_001475&quot;</span>, <span class="st">&quot;NC_002640&quot;</span>)
<span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(acc)){
  ef &lt;-<span class="st"> </span><span class="kw">efetch</span>(acc[i], <span class="st">&quot;nuccore&quot;</span>, <span class="dt">retmode =</span> <span class="st">&quot;text&quot;</span>, <span class="dt">rettype =</span> <span class="st">&quot;gb&quot;</span>)  
  <span class="kw">write</span>(<span class="kw">content</span>(ef),<span class="dt">file=</span><span class="kw">paste</span>(acc[i], <span class="st">&quot;.gb&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;&quot;</span>))
  <span class="kw">Sys.sleep</span>(<span class="dv">1</span>)
  <span class="kw">cat</span>(i, <span class="st">&quot;/&quot;</span>, <span class="kw">length</span>(acc), <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>);<span class="kw">flush.console</span>()
}</code></pre></div>
<div id="genbank-data-parsing" class="section level2">
<h2><span class="header-section-number">12.1</span> Genbank data parsing</h2>
<p>다운로드 받은 파일을 열어 Genbank 포맷의 데이터를 보면 다음과 같이 Locus, Definition 에서부터 시작하여 Feature 까지 다양한 서열에 관한 정보들이 같이 들어 있습니다. 그러나 이러한 정보는 단지 text 파일에 나열된 문자열에 불과합니다. 필요한 정보를 특정 변수에 넣고 이를 필요할 때 언제든 참조해서 사용할 수 있도록 해 두는 것이 좋으며 이러한 작업을 parsing이라고 합니다.</p>
<div class="figure">
<img src="genbankview.png" alt="genbank" />
<p class="caption">genbank</p>
</div>
<p>Genbank 데이터의 파싱은 <code>genbankr</code>이라는 패키지를 사용해서 구현할 수 있습니다. 다음 코드를 이용해서 다운로드받은 첫 번째 뎅기바이러스의 genbank파일을 파싱할 수 있습니다.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(genbankr)

acc &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;NC_001477&quot;</span>, <span class="st">&quot;NC_001474&quot;</span>, <span class="st">&quot;NC_001475&quot;</span>, <span class="st">&quot;NC_002640&quot;</span>)
acc_files &lt;-<span class="st"> </span><span class="kw">paste</span>(acc, <span class="st">&quot;.gb&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;&quot;</span>)

dg_list &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">&quot;list&quot;</span>, <span class="kw">length</span>(acc_files))
dg_list[[<span class="dv">1</span>]] &lt;-<span class="st"> </span><span class="kw">parseGenBank</span>(<span class="dt">file =</span> acc_files[<span class="dv">1</span>])</code></pre></div>
<pre><code>## Warning in fill_stack_df(resthang$FEATURES[typs == &quot;source&quot;]): Got
## unexpected multi-value field(s) [ type ]. The resulting column(s) will
## be of class CharacterList, rather than vector(s). Please contact the
## maintainer if multi-valuedness is expected/meaningful for the listed
## field(s).</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(dg_list[[<span class="dv">1</span>]])</code></pre></div>
<pre><code>## List of 11
##  $ LOCUS     : chr &quot;LOCUS       NC_001477              10735 bp ss-RNA     linear   VRL 03-MAY-2019&quot;
##  $ FEATURES  :List of 37
##   ..$ 1 :&#39;data.frame&#39;:   1 obs. of  11 variables:
##   .. ..$ seqnames: chr &quot;Dengue virus 1&quot;
##   .. ..$ start   : int 1
##   .. ..$ end     : int 10735
##   .. ..$ strand  : chr &quot;+&quot;
##   .. ..$ type    : chr &quot;source&quot;
##   .. ..$ organism: chr &quot;Dengue virus 1&quot;
##   .. ..$ mol_type: chr &quot;genomic RNA&quot;
##   .. ..$ db_xref : chr &quot;taxon:11053&quot;
##   .. ..$ clone   : chr &quot;45AZ5&quot;
##   .. ..$ type.1  : chr &quot;1&quot;
##   .. ..$ loctype : chr &quot;normal&quot;
##   ..$ 2 :&#39;data.frame&#39;:   1 obs. of  6 variables:
##   .. ..$ seqnames: chr &quot;Dengue virus 1&quot;
##   .. ..$ start   : int 1
##   .. ..$ end     : int 94
##   .. ..$ strand  : chr &quot;+&quot;
##   .. ..$ type    : chr &quot;5&#39;UTR&quot;
##   .. ..$ loctype : chr &quot;normal&quot;
##   ..$ 3 :&#39;data.frame&#39;:   1 obs. of  7 variables:
##   .. ..$ seqnames: chr &quot;Dengue virus 1&quot;
##   .. ..$ start   : int 2
##   .. ..$ end     : int 69
##   .. ..$ strand  : chr &quot;+&quot;
##   .. ..$ type    : chr &quot;stem_loop&quot;
##   .. ..$ note    : chr &quot;stem-loop A (SLA)&quot;
##   .. ..$ loctype : chr &quot;normal&quot;
##   ..$ 4 :&#39;data.frame&#39;:   1 obs. of  8 variables:
##   .. ..$ seqnames        : chr &quot;Dengue virus 1&quot;
##   .. ..$ start           : int 70
##   .. ..$ end             : int 78
##   .. ..$ strand          : chr &quot;+&quot;
##   .. ..$ type            : chr &quot;regulatory&quot;
##   .. ..$ regulatory_class: chr &quot;other&quot;
##   .. ..$ note            : chr &quot;oligo U track spacer&quot;
##   .. ..$ loctype         : chr &quot;normal&quot;
##   ..$ 5 :&#39;data.frame&#39;:   1 obs. of  8 variables:
##   .. ..$ seqnames        : chr &quot;Dengue virus 1&quot;
##   .. ..$ start           : int 79
##   .. ..$ end             : int 94
##   .. ..$ strand          : chr &quot;+&quot;
##   .. ..$ type            : chr &quot;regulatory&quot;
##   .. ..$ regulatory_class: chr &quot;promoter&quot;
##   .. ..$ note            : chr &quot;5&#39; upstream AUG region (UAR)&quot;
##   .. ..$ loctype         : chr &quot;normal&quot;
##   ..$ 6 :&#39;data.frame&#39;:   1 obs. of  7 variables:
##   .. ..$ seqnames: chr &quot;Dengue virus 1&quot;
##   .. ..$ start   : int 79
##   .. ..$ end     : int 93
##   .. ..$ strand  : chr &quot;+&quot;
##   .. ..$ type    : chr &quot;stem_loop&quot;
##   .. ..$ note    : chr &quot;stem-loop B (SLB)&quot;
##   .. ..$ loctype : chr &quot;normal&quot;
##   ..$ 7 :&#39;data.frame&#39;:   1 obs. of  10 variables:
##   .. ..$ seqnames    : chr &quot;Dengue virus 1&quot;
##   .. ..$ start       : int 95
##   .. ..$ end         : int 10273
##   .. ..$ strand      : chr &quot;+&quot;
##   .. ..$ type        : chr &quot;gene&quot;
##   .. ..$ gene        : chr &quot;POLY&quot;
##   .. ..$ locus_tag   : chr &quot;DV1_gp1&quot;
##   .. ..$ gene_synonym: chr &quot;polyprotein gene&quot;
##   .. ..$ db_xref     : chr &quot;GeneID:5075725&quot;
##   .. ..$ loctype     : chr &quot;normal&quot;
##   ..$ 8 :&#39;data.frame&#39;:   1 obs. of  14 variables:
##   .. ..$ seqnames    : chr &quot;Dengue virus 1&quot;
##   .. ..$ start       : int 95
##   .. ..$ end         : int 10273
##   .. ..$ strand      : chr &quot;+&quot;
##   .. ..$ type        : chr &quot;CDS&quot;
##   .. ..$ gene        : chr &quot;POLY&quot;
##   .. ..$ locus_tag   : chr &quot;DV1_gp1&quot;
##   .. ..$ gene_synonym: chr &quot;polyprotein gene&quot;
##   .. ..$ codon_start : num 1
##   .. ..$ product     : chr &quot;polyprotein&quot;
##   .. ..$ protein_id  : chr &quot;NP_059433.1&quot;
##   .. ..$ db_xref     : chr &quot;GeneID:5075725&quot;
##   .. ..$ translation : chr &quot;MNNQRKKTGRPSFNMLKRARNRVSTVSQLAKRFSKGLLSGQGPMKLVMAFIAFLRFLAIPPTAGILARWGSFKKNGAIKVLRGFKKEISNMLNIMNRRKRSVTMLLMLLPT&quot;| __truncated__
##   .. ..$ loctype     : chr &quot;normal&quot;
##   ..$ 9 :&#39;data.frame&#39;:   1 obs. of  12 variables:
##   .. ..$ seqnames    : chr &quot;Dengue virus 1&quot;
##   .. ..$ start       : int 95
##   .. ..$ end         : int 436
##   .. ..$ strand      : chr &quot;+&quot;
##   .. ..$ type        : chr &quot;mat_peptide&quot;
##   .. ..$ gene        : chr &quot;POLY&quot;
##   .. ..$ locus_tag   : chr &quot;DV1_gp1&quot;
##   .. ..$ gene_synonym: chr &quot;polyprotein gene&quot;
##   .. ..$ product     : chr &quot;anchored capsid protein ancC&quot;
##   .. ..$ protein_id  : chr &quot;NP_722457.2&quot;
##   .. ..$ db_xref     : chr &quot;VBRC:35735&quot;
##   .. ..$ loctype     : chr &quot;normal&quot;
##   ..$ 10:&#39;data.frame&#39;:   1 obs. of  13 variables:
##   .. ..$ seqnames    : chr &quot;Dengue virus 1&quot;
##   .. ..$ start       : int 95
##   .. ..$ end         : int 394
##   .. ..$ strand      : chr &quot;+&quot;
##   .. ..$ type        : chr &quot;mat_peptide&quot;
##   .. ..$ gene        : chr &quot;POLY&quot;
##   .. ..$ locus_tag   : chr &quot;DV1_gp1&quot;
##   .. ..$ gene_synonym: chr &quot;polyprotein gene&quot;
##   .. ..$ product     : chr &quot;capsid protein C&quot;
##   .. ..$ note        : chr &quot;added by NCBI staff following more recentannotations of this virus sequence&quot;
##   .. ..$ protein_id  : chr &quot;NP_722466.2&quot;
##   .. ..$ db_xref     : chr &quot;VBRC:67793&quot;
##   .. ..$ loctype     : chr &quot;normal&quot;
##   ..$ 11:&#39;data.frame&#39;:   1 obs. of  11 variables:
##   .. ..$ seqnames    : chr &quot;Dengue virus 1&quot;
##   .. ..$ start       : int 437
##   .. ..$ end         : int 934
##   .. ..$ strand      : chr &quot;+&quot;
##   .. ..$ type        : chr &quot;mat_peptide&quot;
##   .. ..$ gene        : chr &quot;POLY&quot;
##   .. ..$ locus_tag   : chr &quot;DV1_gp1&quot;
##   .. ..$ gene_synonym: chr &quot;polyprotein gene&quot;
##   .. ..$ product     : chr &quot;membrane glycoprotein precursor prM&quot;
##   .. ..$ protein_id  : chr &quot;NP_733807.2&quot;
##   .. ..$ loctype     : chr &quot;normal&quot;
##   ..$ 12:&#39;data.frame&#39;:   1 obs. of  12 variables:
##   .. ..$ seqnames    : chr &quot;Dengue virus 1&quot;
##   .. ..$ start       : int 437
##   .. ..$ end         : int 709
##   .. ..$ strand      : chr &quot;+&quot;
##   .. ..$ type        : chr &quot;mat_peptide&quot;
##   .. ..$ gene        : chr &quot;POLY&quot;
##   .. ..$ locus_tag   : chr &quot;DV1_gp1&quot;
##   .. ..$ gene_synonym: chr &quot;polyprotein gene&quot;
##   .. ..$ product     : chr &quot;protein pr&quot;
##   .. ..$ note        : chr &quot;peptide pr&quot;
##   .. ..$ protein_id  : chr &quot;YP_009164956.1&quot;
##   .. ..$ loctype     : chr &quot;normal&quot;
##   ..$ 13:&#39;data.frame&#39;:   1 obs. of  11 variables:
##   .. ..$ seqnames    : chr &quot;Dengue virus 1&quot;
##   .. ..$ start       : int 710
##   .. ..$ end         : int 934
##   .. ..$ strand      : chr &quot;+&quot;
##   .. ..$ type        : chr &quot;mat_peptide&quot;
##   .. ..$ gene        : chr &quot;POLY&quot;
##   .. ..$ locus_tag   : chr &quot;DV1_gp1&quot;
##   .. ..$ gene_synonym: chr &quot;polyprotein gene&quot;
##   .. ..$ product     : chr &quot;membrane glycoprotein M&quot;
##   .. ..$ protein_id  : chr &quot;NP_722459.2&quot;
##   .. ..$ loctype     : chr &quot;normal&quot;
##   ..$ 14:&#39;data.frame&#39;:   1 obs. of  11 variables:
##   .. ..$ seqnames    : chr &quot;Dengue virus 1&quot;
##   .. ..$ start       : int 935
##   .. ..$ end         : int 2419
##   .. ..$ strand      : chr &quot;+&quot;
##   .. ..$ type        : chr &quot;mat_peptide&quot;
##   .. ..$ gene        : chr &quot;POLY&quot;
##   .. ..$ locus_tag   : chr &quot;DV1_gp1&quot;
##   .. ..$ gene_synonym: chr &quot;polyprotein gene&quot;
##   .. ..$ product     : chr &quot;envelope protein E&quot;
##   .. ..$ protein_id  : chr &quot;NP_722460.2&quot;
##   .. ..$ loctype     : chr &quot;normal&quot;
##   ..$ 15:&#39;data.frame&#39;:   1 obs. of  12 variables:
##   .. ..$ seqnames    : chr &quot;Dengue virus 1&quot;
##   .. ..$ start       : int 2420
##   .. ..$ end         : int 3475
##   .. ..$ strand      : chr &quot;+&quot;
##   .. ..$ type        : chr &quot;mat_peptide&quot;
##   .. ..$ gene        : chr &quot;POLY&quot;
##   .. ..$ locus_tag   : chr &quot;DV1_gp1&quot;
##   .. ..$ gene_synonym: chr &quot;polyprotein gene&quot;
##   .. ..$ product     : chr &quot;nonstructural protein NS1&quot;
##   .. ..$ protein_id  : chr &quot;NP_722461.1&quot;
##   .. ..$ db_xref     : chr &quot;VBRC:35739&quot;
##   .. ..$ loctype     : chr &quot;normal&quot;
##   ..$ 16:&#39;data.frame&#39;:   1 obs. of  12 variables:
##   .. ..$ seqnames    : chr &quot;Dengue virus 1&quot;
##   .. ..$ start       : int 3476
##   .. ..$ end         : int 4129
##   .. ..$ strand      : chr &quot;+&quot;
##   .. ..$ type        : chr &quot;mat_peptide&quot;
##   .. ..$ gene        : chr &quot;POLY&quot;
##   .. ..$ locus_tag   : chr &quot;DV1_gp1&quot;
##   .. ..$ gene_synonym: chr &quot;polyprotein gene&quot;
##   .. ..$ product     : chr &quot;nonstructural protein NS2A&quot;
##   .. ..$ protein_id  : chr &quot;NP_733808.1&quot;
##   .. ..$ db_xref     : chr &quot;VBRC:35740&quot;
##   .. ..$ loctype     : chr &quot;normal&quot;
##   ..$ 17:&#39;data.frame&#39;:   1 obs. of  12 variables:
##   .. ..$ seqnames    : chr &quot;Dengue virus 1&quot;
##   .. ..$ start       : int 4130
##   .. ..$ end         : int 4519
##   .. ..$ strand      : chr &quot;+&quot;
##   .. ..$ type        : chr &quot;mat_peptide&quot;
##   .. ..$ gene        : chr &quot;POLY&quot;
##   .. ..$ locus_tag   : chr &quot;DV1_gp1&quot;
##   .. ..$ gene_synonym: chr &quot;polyprotein gene&quot;
##   .. ..$ product     : chr &quot;nonstructural protein NS2B&quot;
##   .. ..$ protein_id  : chr &quot;NP_733809.1&quot;
##   .. ..$ db_xref     : chr &quot;VBRC:35741&quot;
##   .. ..$ loctype     : chr &quot;normal&quot;
##   ..$ 18:&#39;data.frame&#39;:   1 obs. of  12 variables:
##   .. ..$ seqnames    : chr &quot;Dengue virus 1&quot;
##   .. ..$ start       : int 4520
##   .. ..$ end         : int 6376
##   .. ..$ strand      : chr &quot;+&quot;
##   .. ..$ type        : chr &quot;mat_peptide&quot;
##   .. ..$ gene        : chr &quot;POLY&quot;
##   .. ..$ locus_tag   : chr &quot;DV1_gp1&quot;
##   .. ..$ gene_synonym: chr &quot;polyprotein gene&quot;
##   .. ..$ product     : chr &quot;nonstructural protein NS3&quot;
##   .. ..$ protein_id  : chr &quot;NP_722463.1&quot;
##   .. ..$ db_xref     : chr &quot;VBRC:35742&quot;
##   .. ..$ loctype     : chr &quot;normal&quot;
##   ..$ 19:&#39;data.frame&#39;:   1 obs. of  12 variables:
##   .. ..$ seqnames    : chr &quot;Dengue virus 1&quot;
##   .. ..$ start       : int 6377
##   .. ..$ end         : int 6757
##   .. ..$ strand      : chr &quot;+&quot;
##   .. ..$ type        : chr &quot;mat_peptide&quot;
##   .. ..$ gene        : chr &quot;POLY&quot;
##   .. ..$ locus_tag   : chr &quot;DV1_gp1&quot;
##   .. ..$ gene_synonym: chr &quot;polyprotein gene&quot;
##   .. ..$ product     : chr &quot;nonstructural protein NS4A&quot;
##   .. ..$ protein_id  : chr &quot;NP_733810.1&quot;
##   .. ..$ db_xref     : chr &quot;VBRC:35743&quot;
##   .. ..$ loctype     : chr &quot;normal&quot;
##   ..$ 20:&#39;data.frame&#39;:   1 obs. of  13 variables:
##   .. ..$ seqnames    : chr &quot;Dengue virus 1&quot;
##   .. ..$ start       : int 6758
##   .. ..$ end         : int 6826
##   .. ..$ strand      : chr &quot;+&quot;
##   .. ..$ type        : chr &quot;mat_peptide&quot;
##   .. ..$ gene        : chr &quot;POLY&quot;
##   .. ..$ locus_tag   : chr &quot;DV1_gp1&quot;
##   .. ..$ gene_synonym: chr &quot;polyprotein gene&quot;
##   .. ..$ product     : chr &quot;protein 2K&quot;
##   .. ..$ note        : chr &quot;added by NCBI staff following more recentannotations of this virus sequence&quot;
##   .. ..$ protein_id  : chr &quot;NP_722467.1&quot;
##   .. ..$ db_xref     : chr &quot;VBRC:67794&quot;
##   .. ..$ loctype     : chr &quot;normal&quot;
##   ..$ 21:&#39;data.frame&#39;:   1 obs. of  12 variables:
##   .. ..$ seqnames    : chr &quot;Dengue virus 1&quot;
##   .. ..$ start       : int 6827
##   .. ..$ end         : int 7573
##   .. ..$ strand      : chr &quot;+&quot;
##   .. ..$ type        : chr &quot;mat_peptide&quot;
##   .. ..$ gene        : chr &quot;POLY&quot;
##   .. ..$ locus_tag   : chr &quot;DV1_gp1&quot;
##   .. ..$ gene_synonym: chr &quot;polyprotein gene&quot;
##   .. ..$ product     : chr &quot;nonstructural protein NS4B&quot;
##   .. ..$ protein_id  : chr &quot;NP_733811.1&quot;
##   .. ..$ db_xref     : chr &quot;VBRC:35744&quot;
##   .. ..$ loctype     : chr &quot;normal&quot;
##   ..$ 22:&#39;data.frame&#39;:   1 obs. of  12 variables:
##   .. ..$ seqnames    : chr &quot;Dengue virus 1&quot;
##   .. ..$ start       : int 7574
##   .. ..$ end         : int 10270
##   .. ..$ strand      : chr &quot;+&quot;
##   .. ..$ type        : chr &quot;mat_peptide&quot;
##   .. ..$ gene        : chr &quot;POLY&quot;
##   .. ..$ locus_tag   : chr &quot;DV1_gp1&quot;
##   .. ..$ gene_synonym: chr &quot;polyprotein gene&quot;
##   .. ..$ product     : chr &quot;RNA-dependent RNA polymerase NS5&quot;
##   .. ..$ protein_id  : chr &quot;NP_722465.1&quot;
##   .. ..$ db_xref     : chr &quot;VBRC:35745&quot;
##   .. ..$ loctype     : chr &quot;normal&quot;
##   ..$ 23:&#39;data.frame&#39;:   1 obs. of  7 variables:
##   .. ..$ seqnames: chr &quot;Dengue virus 1&quot;
##   .. ..$ start   : int 114
##   .. ..$ end     : int 130
##   .. ..$ strand  : chr &quot;+&quot;
##   .. ..$ type    : chr &quot;stem_loop&quot;
##   .. ..$ note    : chr &quot;capsid region hairpin (cHP)&quot;
##   .. ..$ loctype : chr &quot;normal&quot;
##   ..$ 24:&#39;data.frame&#39;:   1 obs. of  8 variables:
##   .. ..$ seqnames        : chr &quot;Dengue virus 1&quot;
##   .. ..$ start           : int 132
##   .. ..$ end             : int 142
##   .. ..$ strand          : chr &quot;+&quot;
##   .. ..$ type            : chr &quot;regulatory&quot;
##   .. ..$ regulatory_class: chr &quot;other&quot;
##   .. ..$ note            : chr &quot;5&#39; conserved sequence (CS); also called cyclizationsequence&quot;
##   .. ..$ loctype         : chr &quot;normal&quot;
##   ..$ 25:&#39;data.frame&#39;:   1 obs. of  6 variables:
##   .. ..$ seqnames: chr &quot;Dengue virus 1&quot;
##   .. ..$ start   : int 10274
##   .. ..$ end     : int 10735
##   .. ..$ strand  : chr &quot;+&quot;
##   .. ..$ type    : chr &quot;3&#39;UTR&quot;
##   .. ..$ loctype : chr &quot;normal&quot;
##   ..$ 26:&#39;data.frame&#39;:   1 obs. of  9 variables:
##   .. ..$ seqnames   : chr &quot;Dengue virus 1&quot;
##   .. ..$ start      : int 10320
##   .. ..$ end        : int 10735
##   .. ..$ strand     : chr &quot;+&quot;
##   .. ..$ type       : chr &quot;ncRNA&quot;
##   .. ..$ ncRNA_class: chr &quot;lncRNA&quot;
##   .. ..$ product    : chr &quot;sfRNA1&quot;
##   .. ..$ note       : chr &quot;subgenomic flavivirus RNA&quot;
##   .. ..$ loctype    : chr &quot;normal&quot;
##   ..$ 27:&#39;data.frame&#39;:   1 obs. of  7 variables:
##   .. ..$ seqnames: chr &quot;Dengue virus 1&quot;
##   .. ..$ start   : int 10324
##   .. ..$ end     : int 10389
##   .. ..$ strand  : chr &quot;+&quot;
##   .. ..$ type    : chr &quot;stem_loop&quot;
##   .. ..$ note    : chr &quot;flaviviral nuclease-resistant RNA 1 (fNR1); alsocalled stem-loop 1 or xrRNA1&quot;
##   .. ..$ loctype : chr &quot;normal&quot;
##   ..$ 28:&#39;data.frame&#39;:   1 obs. of  9 variables:
##   .. ..$ seqnames   : chr &quot;Dengue virus 1&quot;
##   .. ..$ start      : int 10393
##   .. ..$ end        : int 10735
##   .. ..$ strand     : chr &quot;+&quot;
##   .. ..$ type       : chr &quot;ncRNA&quot;
##   .. ..$ ncRNA_class: chr &quot;lncRNA&quot;
##   .. ..$ product    : chr &quot;sfRNA2&quot;
##   .. ..$ note       : chr &quot;subgenomic flavivirus RNA&quot;
##   .. ..$ loctype    : chr &quot;normal&quot;
##   ..$ 29:&#39;data.frame&#39;:   1 obs. of  7 variables:
##   .. ..$ seqnames: chr &quot;Dengue virus 1&quot;
##   .. ..$ start   : int 10397
##   .. ..$ end     : int 10458
##   .. ..$ strand  : chr &quot;+&quot;
##   .. ..$ type    : chr &quot;stem_loop&quot;
##   .. ..$ note    : chr &quot;flaviviral nuclease-resistant RNA 2 (fNR2); alsocalled stem-loop 2 or xrRNA2&quot;
##   .. ..$ loctype : chr &quot;normal&quot;
##   ..$ 30:&#39;data.frame&#39;:   1 obs. of  9 variables:
##   .. ..$ seqnames   : chr &quot;Dengue virus 1&quot;
##   .. ..$ start      : int 10463
##   .. ..$ end        : int 10735
##   .. ..$ strand     : chr &quot;+&quot;
##   .. ..$ type       : chr &quot;ncRNA&quot;
##   .. ..$ ncRNA_class: chr &quot;lncRNA&quot;
##   .. ..$ product    : chr &quot;sfRNA3&quot;
##   .. ..$ note       : chr &quot;subgenomic flavivirus RNA&quot;
##   .. ..$ loctype    : chr &quot;normal&quot;
##   ..$ 31:&#39;data.frame&#39;:   1 obs. of  7 variables:
##   .. ..$ seqnames: chr &quot;Dengue virus 1&quot;
##   .. ..$ start   : int 10467
##   .. ..$ end     : int 10549
##   .. ..$ strand  : chr &quot;+&quot;
##   .. ..$ type    : chr &quot;stem_loop&quot;
##   .. ..$ note    : chr &quot;dumbbell 1 (DBI); also called xrRNA3&quot;
##   .. ..$ loctype : chr &quot;normal&quot;
##   ..$ 32:&#39;data.frame&#39;:   1 obs. of  9 variables:
##   .. ..$ seqnames   : chr &quot;Dengue virus 1&quot;
##   .. ..$ start      : int 10551
##   .. ..$ end        : int 10735
##   .. ..$ strand     : chr &quot;+&quot;
##   .. ..$ type       : chr &quot;ncRNA&quot;
##   .. ..$ ncRNA_class: chr &quot;lncRNA&quot;
##   .. ..$ product    : chr &quot;sfRNA4&quot;
##   .. ..$ note       : chr &quot;subgenomic flavivirus RNA&quot;
##   .. ..$ loctype    : chr &quot;normal&quot;
##   ..$ 33:&#39;data.frame&#39;:   1 obs. of  7 variables:
##   .. ..$ seqnames: chr &quot;Dengue virus 1&quot;
##   .. ..$ start   : int 10551
##   .. ..$ end     : int 10633
##   .. ..$ strand  : chr &quot;+&quot;
##   .. ..$ type    : chr &quot;stem_loop&quot;
##   .. ..$ note    : chr &quot;dumbbell 2 (DBII); also called xrRNA4&quot;
##   .. ..$ loctype : chr &quot;normal&quot;
##   ..$ 34:&#39;data.frame&#39;:   1 obs. of  8 variables:
##   .. ..$ seqnames        : chr &quot;Dengue virus 1&quot;
##   .. ..$ start           : int 10630
##   .. ..$ end             : int 10640
##   .. ..$ strand          : chr &quot;+&quot;
##   .. ..$ type            : chr &quot;regulatory&quot;
##   .. ..$ regulatory_class: chr &quot;other&quot;
##   .. ..$ note            : chr &quot;3&#39; conserved sequence (CS); also called cyclizationsequence&quot;
##   .. ..$ loctype         : chr &quot;normal&quot;
##   ..$ 35:&#39;data.frame&#39;:   1 obs. of  7 variables:
##   .. ..$ seqnames: chr &quot;Dengue virus 1&quot;
##   .. ..$ start   : int 10643
##   .. ..$ end     : int 10656
##   .. ..$ strand  : chr &quot;+&quot;
##   .. ..$ type    : chr &quot;stem_loop&quot;
##   .. ..$ note    : chr &quot;short hairpin (sHP)&quot;
##   .. ..$ loctype : chr &quot;normal&quot;
##   ..$ 36:&#39;data.frame&#39;:   1 obs. of  8 variables:
##   .. ..$ seqnames        : chr &quot;Dengue virus 1&quot;
##   .. ..$ start           : int 10654
##   .. ..$ end             : int 10670
##   .. ..$ strand          : chr &quot;+&quot;
##   .. ..$ type            : chr &quot;regulatory&quot;
##   .. ..$ regulatory_class: chr &quot;promoter&quot;
##   .. ..$ note            : chr &quot;3&#39; upstream AUG region (UAR)&quot;
##   .. ..$ loctype         : chr &quot;normal&quot;
##   ..$ 37:&#39;data.frame&#39;:   1 obs. of  7 variables:
##   .. ..$ seqnames: chr &quot;Dengue virus 1&quot;
##   .. ..$ start   : int 10657
##   .. ..$ end     : int 10735
##   .. ..$ strand  : chr &quot;+&quot;
##   .. ..$ type    : chr &quot;stem_loop&quot;
##   .. ..$ note    : chr &quot;3&#39; stem-loop (3&#39;SL)&quot;
##   .. ..$ loctype : chr &quot;normal&quot;
##   ..- attr(*, &quot;dim&quot;)= int 37
##   ..- attr(*, &quot;dimnames&quot;)=List of 1
##   .. ..$ : chr [1:37] &quot;1&quot; &quot;2&quot; &quot;3&quot; &quot;4&quot; ...
##  $ ORIGIN    :Formal class &#39;DNAStringSet&#39; [package &quot;Biostrings&quot;] with 5 slots
##   .. ..@ pool           :Formal class &#39;SharedRaw_Pool&#39; [package &quot;XVector&quot;] with 2 slots
##   .. .. .. ..@ xp_list                    :List of 1
##   .. .. .. .. ..$ :&lt;externalptr&gt; 
##   .. .. .. ..@ .link_to_cached_object_list:List of 1
##   .. .. .. .. ..$ :&lt;environment: 0x000000001f77c820&gt; 
##   .. ..@ ranges         :Formal class &#39;GroupedIRanges&#39; [package &quot;XVector&quot;] with 7 slots
##   .. .. .. ..@ group          : int 1
##   .. .. .. ..@ start          : int 1
##   .. .. .. ..@ width          : int 10735
##   .. .. .. ..@ NAMES          : chr &quot;Dengue virus 1&quot;
##   .. .. .. ..@ elementType    : chr &quot;ANY&quot;
##   .. .. .. ..@ elementMetadata: NULL
##   .. .. .. ..@ metadata       : list()
##   .. ..@ elementType    : chr &quot;DNAString&quot;
##   .. ..@ elementMetadata: NULL
##   .. ..@ metadata       : list()
##  $ ACCESSION : chr &quot;NC_001477&quot;
##  $ COMMENT   : NULL
##  $ DBLINK    : NULL
##  $ DEFINITION: chr &quot;Dengue virus 1, complete genome.&quot;
##  $ KEYWORDS  : chr &quot;RefSeq.&quot;
##  $ REFERENCE : NULL
##  $ SOURCE    :List of 3
##   ..$ source  : chr &quot;Dengue virus 1&quot;
##   ..$ organism: chr &quot;Dengue virus 1&quot;
##   ..$ lineage : chr [1:4] &quot;            Viruses&quot; &quot;Riboviria&quot; &quot;Flaviviridae&quot; &quot;Flavivirus.&quot;
##  $ VERSION   : Named chr [1:2] &quot;NC_001477.1&quot; NA
##   ..- attr(*, &quot;names&quot;)= chr [1:2] &quot;accession.version&quot; &quot;GenInfoID&quot;</code></pre>
<p>위 파싱된 파일의 구조를 보면 37개의 feature가 있고 각 feature에 대해서 7 또는 8개의 variable로 이루어진 것을 알 수 있습니다. 즉, dg_list가 4개 길이를 갖는 list형태의 변수이고 4개 각각의 저장된 변수가 다시 또 list 형태의 변수이고 그 안에 여러개의 feature가 data.frame 형식의 변수로 저장이 되어 있는 형태 입니다. 예를 들어 37번째 feature의 start position을 알고 싶을 때 <code>dg_list[[1]]$FEATURES[[37]]$start</code> 이와 같이 실행하면 원하는 값이 출력됩니다.</p>
<div id="exercise-11-1" class="section level3">
<h3><span class="header-section-number">12.1.1</span> Exercise 11-1</h3>
<p>앞서 파싱 코드는 첫 번째 파일에 대해서만 수행되었습니다. 이를 <code>for</code>문을 이용하여 4개 파일 모두에 대해서 파싱 작업을 수행합니다. 프로그램밍에서 <code>for</code>문의 사용은 R 언어뿐만 아니라 다른 여러 언어에서도 기본이 되는 문법으로 그 사용법은 필수로 숙지하고 있어야 합니다. for 문과 i 인덱스를 사용하는 부분을 특히 참고하시기 바랍니다.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">acc &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;NC_001477&quot;</span>, <span class="st">&quot;NC_001474&quot;</span>, <span class="st">&quot;NC_001475&quot;</span>, <span class="st">&quot;NC_002640&quot;</span>)
acc_files &lt;-<span class="st"> </span><span class="kw">paste</span>(acc, <span class="st">&quot;.gb&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;&quot;</span>)

dg_list &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">&quot;list&quot;</span>, <span class="kw">length</span>(acc_files))
<span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(acc)){
  dg_list[[i]] &lt;-<span class="st"> </span><span class="kw">parseGenBank</span>(<span class="dt">file =</span> acc_files[i])  
}</code></pre></div>
</div>
</div>
<div id="iranges-for-genome-view" class="section level2">
<h2><span class="header-section-number">12.2</span> IRanges for genome view</h2>
<p>앞서 보여드린 바와 같이 genbank 데이터의 내용 중 대부분을 차지하는 정보는 전체 지놈 서열 중 어디서 어디까지가 유전자 또는 coding sequence 이고 그 번역된 정보가 무엇인지 설명하는 정보 입니다. 즉, feature 정보가 genbank 데이터의 대부분을 차지한다고 해도 과언이 아니며 이러한 feature 정보를 보는 것이 곧 지놈을 분석하기 위한 필수 조건이라고 해도 과언이 아닙니다. 따라서 genbank의 feature 정보를 효율적으로 가시화 하기위한 방법들이 다양하게 개발되어 왔으며 R에서는 IRanges와 GenomicRanges 라는 패키지를 사용할 수 있습니다.</p>
<div id="exercise-11-2" class="section level3">
<h3><span class="header-section-number">12.2.1</span> Exercise 11-2</h3>
<p>지놈 가시화에서 각 구간(Feature)에 해당하는 CDS 는 사각형으로 그리는 것이 보통입니다. 따라서 IRanges 를 사용하기 전에 사각형을 그리는 예제를 수행하면서 feature를 어떻게 표현할 수 있는이 먼저 이해해 보도록 하겠습니다. 사각형은 기본적으로 4개의 좌표가 있으면 그릴 수 있는 도형입니다.</p>
<div class="figure">
<img src="rect.png" alt="Rect" />
<p class="caption">Rect</p>
</div>
<p>위와 같은 사각형은 기본 R 그래픽 함수를 이용해서 사각형을 그릴 수 있습니다. <code>plot.new()</code>함수는 빈 켄버스를 그리고 <code>plot.window()</code> 함수는 x, y 축의 범위 등을 지정하며 이 후 rect, axis 등을 이용해서 사각형이나 x, y 축을 그려줍니다.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot.new</span>()
<span class="kw">plot.window</span>(<span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">2</span>), <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">2</span>))
<span class="kw">rect</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">1</span>)
<span class="kw">axis</span>(<span class="dv">1</span>)
<span class="kw">axis</span>(<span class="dv">2</span>)</code></pre></div>
<p><img src="Rprog_files/figure-html/unnamed-chunk-165-1.png" width="672" /></p>
<p>이러한 사각형의 연속된 표현으로 지놈상의 CDS를 표현해 주며 사각형 정보를 담고 있는 변수는 IRanges를 이용해서 만들 수 있습니다. IRanges 함수는 start와 end를 매개변수로 받으며 각각 vector 형으로 vector 길이 만큼의 start, end 짝이 생성되어 하나의 변수에 저장됩니다. 다음 코드를 실행하면서 좀 더 쉽게 이해 하실 수 있습니다.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(IRanges)

ir &lt;-<span class="st"> </span><span class="kw">IRanges</span>(<span class="dt">start=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">12</span>,<span class="dv">10</span>), <span class="dt">end=</span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">25</span>,<span class="dv">19</span>))
ir
<span class="kw">length</span>(ir)
<span class="kw">start</span>(ir)
<span class="kw">end</span>(ir)
<span class="kw">width</span>(ir)
<span class="kw">range</span>(ir)</code></pre></div>
<p>이제 앞서 배운 사각형 그리는 방법을 이용해서 위 IRanges로 만든 각각의 구간을 그려보도록 하겠습니다. 모든 구간은 start 부터 end 까지 주어져 있고 각 구간의 높이를 1로 고정한다면 사각형을 그리기 위한 네 점을 모두 구할 수 있습니다. 한 가지 주의를 요하는 부분은 각 CDS의 구간이 서로 독립적이지 않고 서로 겹쳐져서 나타날 수 있습니다. 이 때 겹치는 부분이 있을 경우 사각형을 한 칸 위로 위치시켜서 그리면 모든 CDS를 명확히 구분되게 그릴 수 있습니다.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ir &lt;-<span class="st"> </span><span class="kw">IRanges</span>(<span class="dt">start=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">12</span>,<span class="dv">10</span>), <span class="dt">end=</span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">25</span>,<span class="dv">19</span>))

height &lt;-<span class="st"> </span><span class="dv">1</span> 
xlim =<span class="st"> </span><span class="kw">c</span>(<span class="kw">min</span>(<span class="kw">start</span>(ir)), <span class="kw">max</span>(<span class="kw">end</span>(ir)))
bins =<span class="st"> </span><span class="kw">disjointBins</span>(ir)  

<span class="kw">plot.new</span>()
<span class="kw">plot.window</span>(xlim, <span class="kw">c</span>(<span class="dv">0</span>, <span class="kw">max</span>(bins) <span class="op">*</span><span class="st"> </span>height))
ybottom =<span class="st"> </span>bins <span class="op">*</span><span class="st"> </span>height <span class="op">-</span><span class="st"> </span>height
<span class="kw">rect</span>(<span class="kw">start</span>(ir), ybottom, <span class="kw">end</span>(ir), ybottom <span class="op">+</span><span class="st"> </span>height, <span class="dt">col =</span> <span class="st">&quot;blue&quot;</span>)
<span class="kw">axis</span>(<span class="dv">1</span>)</code></pre></div>
<p><img src="Rprog_files/figure-html/unnamed-chunk-167-1.png" width="672" /></p>
<p>이제 동일한 기능을 기본 그래프 함수가 아닌 ggplot을 이용해서 수행해 보겠습니다. ggplot은 앞에서 배운 것처럼 plot 정보를 저장한 data.frame형 변수를 만들고 aes로 해당 데이터와 필요 파라메터들을 mapping해준 후 적절한 geometry 함수와 stat을 이용해서 그림을 그리게 됩니다. 본 예제에서는 geom_rect라는 geometry 함수를 사용하면 되겠으며 geom_rect 함수의 필요 파라메터는 사각형을 그리기 위해서 필요한 (xmin, xmax, ymin and ymax) 네 가지 값을 전달해주면 됩니다. 즉, aes 함수에서 이 네 개 파라메터에 맞는 데이터를 mapping 해 주면 되며 이를 위해서 data.frame 또한 위 네 개 값을 갖도록 만들어 주면 되겠습니다. 대략적인 코드는 다음과 같습니다.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ir &lt;-<span class="st"> </span><span class="kw">IRanges</span>(<span class="dt">start=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">12</span>,<span class="dv">10</span>), <span class="dt">end=</span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">25</span>,<span class="dv">19</span>))

height &lt;-<span class="st"> </span><span class="dv">1</span> 
bins &lt;-<span class="st"> </span><span class="kw">disjointBins</span>(ir)  
ybottom &lt;-<span class="st"> </span>bins<span class="op">*</span>height <span class="op">-</span><span class="st"> </span>height
df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">ybottom =</span> ybottom, 
    <span class="dt">xleft =</span> <span class="kw">start</span>(ir), 
    <span class="dt">xright =</span> <span class="kw">end</span>(ir), 
    <span class="dt">ytop =</span> ybottom <span class="op">+</span><span class="st"> </span>height)
<span class="kw">ggplot</span>(df, <span class="kw">aes</span>(<span class="dt">xmax =</span> xright, <span class="dt">xmin =</span> xleft, <span class="dt">ymax =</span> ytop, <span class="dt">ymin =</span> ybottom)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_rect</span>() </code></pre></div>
<p><img src="Rprog_files/figure-html/unnamed-chunk-168-1.png" width="672" /></p>
</div>
<div id="exercise-11-3" class="section level3">
<h3><span class="header-section-number">12.2.2</span> Exercise 11-3</h3>
<p>위와 같이 생성한 IRanges 타입 변수의 정보를 그리기 위해서는 1) df 라는 각 사각형의 위치 정보를 갖는 데이터프레임을 생성하고 2) ggplot을 이용해서 적절히 그려주는 (크게) 두 가지 일을 수행해야 합니다. 그런데 새로운 ir 만들 때마다 위 두 가지를 수행하는 코드를 작성하기 보다는 하나의 함수로 만들어서 사용하면 효율을 높일 수 있습니다. 다음 코드에 위 두 가지 일을 수행하는 <code>plotRanges</code>라는 함수를 만들었습니다.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">plotRanges &lt;-<span class="st"> </span><span class="cf">function</span>(ir, <span class="dt">height=</span><span class="dv">1</span>, <span class="dt">sep=</span><span class="fl">0.5</span>){
  xm &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">min</span>(<span class="kw">start</span>(ir))<span class="op">-</span>sep, <span class="kw">max</span>(<span class="kw">end</span>(ir))<span class="op">+</span>sep)
  bins &lt;-<span class="st"> </span><span class="kw">disjointBins</span>(ir)  
  ybottom &lt;-<span class="st"> </span>bins<span class="op">*</span>(sep<span class="op">+</span>height) <span class="op">-</span><span class="st"> </span>height
  df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">ybottom =</span> ybottom, 
                   <span class="dt">xleft =</span> <span class="kw">start</span>(ir) <span class="op">-</span><span class="st"> </span>sep, 
                   <span class="dt">xright =</span> <span class="kw">end</span>(ir) <span class="op">+</span><span class="st"> </span>sep, 
                   <span class="dt">ytop =</span> ybottom <span class="op">+</span><span class="st"> </span>height)
  <span class="kw">ggplot</span>(df, <span class="kw">aes</span>(<span class="dt">xmax =</span> xright, <span class="dt">xmin =</span> xleft, <span class="dt">ymax =</span> ytop, <span class="dt">ymin =</span> ybottom)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_rect</span>() 
}</code></pre></div>
<p>이제 IRanges 변수를 만든 후 위 함수를 다음과 같이 간단히 사용할 수 있습니다.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ir &lt;-<span class="st"> </span><span class="kw">IRanges</span>(<span class="dt">start=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">12</span>,<span class="dv">10</span>), <span class="dt">end=</span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">25</span>,<span class="dv">19</span>))
<span class="kw">plotRanges</span>(ir)</code></pre></div>
<p><img src="Rprog_files/figure-html/unnamed-chunk-170-1.png" width="672" /></p>
</div>
</div>
<div id="feature-view-using-iranges-and-ggplot2" class="section level2">
<h2><span class="header-section-number">12.3</span> Feature view using IRanges and ggplot2</h2>
<p>이제 앞서 읽어온 실제 뎅기바이러스 지놈 데이터로부터 feature를 읽고 이를 visualization 하는 실습을 하겠습니다. 먼저 리스트에 저장된 네 개의 뎅기 바이러스 genbank 정보를 하나의 변수 (sel_dg)에 옮기고 각 정보들을 확인해 봅니다. 이 후 <code>for</code>문을 이용해서 모든 feature 들에 대해 start와 end 정보를 읽어 옵니다.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sel_dg &lt;-<span class="st"> </span>dg_list[[<span class="dv">1</span>]]

<span class="co">#sel_dg$FEATURES[[1]]$type</span>
<span class="co">#sel_dg$FEATURES[[1]]$start</span>
<span class="co">#sel_dg$FEATURES[[1]]$end</span>
<span class="co">#sel_dg$FEATURES[[1]]$strand</span>
<span class="co">#sel_dg$FEATURES[[1]]$product</span>

n &lt;-<span class="st"> </span><span class="kw">length</span>(sel_dg<span class="op">$</span>FEATURES)
start.pos &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">0</span>, n)
end.pos &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">0</span>, n)

<span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n){
  start.pos[i] &lt;-<span class="st"> </span>sel_dg<span class="op">$</span>FEATURES[[i]]<span class="op">$</span>start
  end.pos[i] &lt;-<span class="st"> </span>sel_dg<span class="op">$</span>FEATURES[[i]]<span class="op">$</span>end
}</code></pre></div>
<p>그리고 위 읽어들인 feature들의 정보를 IRanges 함수로 구간별 정보를 갖는 ir 변수를 만들고 앞서 만든 <code>plotRanges()</code>라는 함수를 이용하여 각 구간을 그려 보도록 하겠습니다.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ir &lt;-<span class="st"> </span><span class="kw">IRanges</span>(<span class="dt">start=</span>start.pos, <span class="dt">end=</span>end.pos)
<span class="kw">plotRanges</span>(ir)</code></pre></div>
<p><img src="Rprog_files/figure-html/unnamed-chunk-172-1.png" width="672" /></p>
<div id="exercise-11-4" class="section level3">
<h3><span class="header-section-number">12.3.1</span> Exercise 11-4</h3>
<p>그런데 위 그림에서는 feature 정보들이 눈에 띄게 들어오지 않습니다. 이는 두 가지 문제가 원인으로 볼 수 있습니다. 하나는 feature들이 같은 색이기 때문에 구분되지 않는 상황과 또 하나는 다양한 type의 feature가 있기 때문에 우리가 보고싶은 feature만 따로 빼내기 전에는 많은 정보들이 혼재되어 보일 수 있다는 점 입니다.</p>
<p>이를 위해 우리는 feature type이 “mat_peptide” (즉, 성숙한 팹타이드?) 만을 골라서 그림을 그려보도록 하겠습니다. 조건문을 실행하기 위해서는 <code>if</code>를 사용할 수 있습니다. 또한 앞서 코드와는 다르게 <code>for</code>문을 반복하면서 vector를 하나씩 늘려가는 방법으로 값을 저장하겠습니다.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sel_dg &lt;-<span class="st"> </span>dg_list[[<span class="dv">1</span>]]
n &lt;-<span class="st"> </span><span class="kw">length</span>(sel_dg<span class="op">$</span>FEATURES)
start_pos &lt;-<span class="st"> </span><span class="kw">c</span>()
end_pos &lt;-<span class="st"> </span><span class="kw">c</span>()

<span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n){
  <span class="cf">if</span>(sel_dg<span class="op">$</span>FEATURES[[i]]<span class="op">$</span>type<span class="op">==</span><span class="st">&quot;mat_peptide&quot;</span>){
    start_pos &lt;-<span class="st"> </span><span class="kw">c</span>(start_pos, sel_dg<span class="op">$</span>FEATURES[[i]]<span class="op">$</span>start)
    end_pos &lt;-<span class="st"> </span><span class="kw">c</span>(end_pos, sel_dg<span class="op">$</span>FEATURES[[i]]<span class="op">$</span>end)
  }
}

ir &lt;-<span class="st"> </span><span class="kw">IRanges</span>(<span class="dt">start=</span>start_pos, <span class="dt">end=</span>end_pos)
<span class="kw">plotRanges</span>(ir)</code></pre></div>
<p><img src="Rprog_files/figure-html/unnamed-chunk-173-1.png" width="672" /></p>
<p>우리가 필요한 feature 만을 뽑아서 본 그림 입니다. 그러나 총 feature 의 갯수는 13개인데 여전히 구분이 되지 않고 있습니다. 이는 앞서 언급한 문제점 중 첫번째 문제점 때문에 발생하는 상황입니다. 이를 위해 다음과 같이 임의로 group을 나누어 주고 각 그룹에 대해서 다른 색을 그리도혹 ggplot의 명령을 바꾸어 주면 기대한 것 처럼 각 feature들 구분할 수 있습니다.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ir &lt;-<span class="st"> </span><span class="kw">IRanges</span>(start_pos, end_pos)
bins &lt;-<span class="st"> </span><span class="kw">disjointBins</span>(ir)
df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">ybottom =</span> bins, 
                 <span class="dt">xleft =</span> <span class="kw">start</span>(ir), 
                 <span class="dt">xright =</span> <span class="kw">end</span>(ir), 
                 <span class="dt">ytop =</span> bins <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, 
                 <span class="dt">group_no =</span> LETTERS[<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(ir)])
<span class="kw">ggplot</span>(df, <span class="kw">aes</span>(<span class="dt">xmax =</span> xright, 
               <span class="dt">xmin =</span> xleft, 
               <span class="dt">ymax =</span> ytop, 
               <span class="dt">ymin =</span> ybottom,
               <span class="dt">fill =</span> group_no)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_rect</span>()</code></pre></div>
<p><img src="Rprog_files/figure-html/unnamed-chunk-174-1.png" width="672" /></p>
<p>IRanges는 특정 CDS 구간에 대한 범위 정보만을 가지고 있는 데이터 타입입니다. 그러나 해당 위치의 locus 또는 유전자 이름이나 단백질 정보 등 해당 서열에 해당하는 많은 유전학적 정보들이 있을 수 있습니다. GenomicRanges는 IRanges를 포함하여 위 정보들을 모두 다룰 수 있는 패키지 입니다.</p>
<p>다음 시간에는 위 GenomicRanges를 간단히만 설명하고 서열 비교 등에 관한 내용을 학습하도록 하겠습니다.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="lecture-10-sequence-analysis-in-r-i.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
